src/vsphere_cpi/lib/cloud/vsphere/cluster_picker.rb

```diff
@@ -67,14 +72,55 @@ def best_cluster_placement(req_memory:, disk_configurations:)
 
     private
 
+    def placements_with_gpu_attached_hosts(placements, clusters, gpu_config)
+      # return if gpu_config is nil or number_of_gpu are nont specified or number of gpu are zero
+      return if gpu_config.nil? || gpu_config[:number_of_gpus].nil? || gpu_config[:number_of_gpus] == 0
+
+      clusters.each do |cluster|
+        placements[cluster.name][:host_gpu_map] = {}
+        cluster.mob.host.each do |host|
+          # Get list of available gpus
+          available_gpus = available_gpus(host, gpu_config[:number_of_gpus])
+          # Add this list to host_gpu_map with host name as the key unless this list is empty
+          placements[cluster.name][:host_gpu_map][host.name] = available_gpus unless available_gpus.empty?
+        end
+        # Remove the placement if it has no such hosts
+        placements.delete(cluster.name) if placements[cluster.name][:host].empty?
+      end
+      placements
+    end
+
+    def available_gpus(host_mob, num_gpu_required)
```

Comment:
isnt this method racy since multiple vms in an instance group could be created in parallel?

Reply:
It is  racy but so is the selection of datastore and cluster. CPI has no control over it since for each create_vm call bosh starts a unique CPI process. 
I can randomize the selection of host(with GPUs) from a cluster like we do for datastores. Beyond that, The only way I see a solution is to add random sleeps in create_vm to avoid races.
@yeshwantbabar : we need to discuss this.

Reply:
This feature is experimental and has few more assumptions in the current form; like, breaking drs affinity rules.
For now, I think best way will be to create a separate vm_type with GPU specs and create limited instances with GPUs in a separate cluster. Since each create_vm will see same number of gpus available on each host of the cluster, it will be best to have enough gpus on each host.